js/src/ipc/writer/binary.ts

```diff
@@ -269,17 +269,17 @@ export class RecordBatchSerializer extends VectorVisitor {
         const byteLength = Math.min(lastOffset - firstOffset, values.byteLength - firstOffset);
         // Push in the order FlatList types read their buffers
         // valueOffsets buffer first
-        this.addBuffer(this.getZeroBasedValueOffsets(offset, length, valueOffsets));
+        this.addBuffer(this.getZeroBasedValueOffsets(0, length, valueOffsets));
         // sliced values buffer second
         this.addBuffer(values.subarray(firstOffset + offset, firstOffset + offset + byteLength));
         return this;
     }
     protected visitListVector<T extends SingleNestedType>(vector: Vector<T>) {
         const { data, length } = vector;
-        const { offset, valueOffsets } = <any> data;
+        const { valueOffsets } = <any> data;
         // If we have valueOffsets (ListVector), push that buffer first
         if (valueOffsets) {
-            this.addBuffer(this.getZeroBasedValueOffsets(offset, length, valueOffsets));
+            this.addBuffer(this.getZeroBasedValueOffsets(0, length, valueOffsets));
```

Comment:
@TheNeuralBit you're right that we `subarray` the valueOffsets buffer when we slice, but we still need to call `getZeroBasedValueOffsets()` in all these places because it also does work in the case that [`valueOffsets[0] != 0`](https://github.com/apache/arrow/blob/bcd58caa59d26f6c15d4eb9f7ded758d4227f10f/js/src/ipc/writer/binary.ts#L324).